crates/sparrow-execution/src/lib.rs

```diff
@@ -8,254 +8,193 @@
-        let mut worker_pool = WorkerPool::start(query_id).change_context(Error::Creating)?;
+    ) -> error_stack::Result<Self, Error> {
+        let mut executor = PlanExecutor {
+            worker_pool: WorkerPool::new(query_id).change_context(Error::Creating)?,
+            source_tasks: SourceTasks::default(),
+        };
 
-        // This sets up some fake stuff:
-        // - We don't have sources / sinks yet, so we use tokio channels.
-        // - We create a "hypothetical" scan step (0)
-        // - We create a hard-coded "project" step (1)
-        // - We output the results to the channel.
+        let sink_pipeline = executor
+            .worker_pool
+            .add_pipeline(1, WriteChannelPipeline::new(output));
+
+        // Map from the producing step ID to the consumers.
+        let mut consumers: HashMap<StepId, PipelineInputs> = HashMap::new();
+        consumers
+            .entry(plan.steps.last_idx())
+            .or_default()
+            .add_consumer(sink_pipeline, 0);
+
+        // Iterate in reverse so the receivers (consumers) are created before the senders (producers).
+        for pipeline in plan.pipelines.iter().rev() {
+            let first_step_id = *pipeline.steps.first().expect("at least one step");
+            let last_step_id = *pipeline.steps.last().expect("at least one step");
+            let sink = consumers
+                .remove(&last_step_id)
+                .expect("at least one consumer for step");
+
+            // This is a limitation of how the `PipelineInput` is identified. Ideally,
+            // any pipeline should be connect-able to multiple down-stream pipelines.
+            debug_assert_eq!(sink.len(), 1, "Only one consumer currently supported");
+
+            let first_step_inputs = &plan.steps[first_step_id].inputs;
+
+            // Create a transform pipeline (if possible), othrewise create the
+            // appropriate non-transform pipeline.
+            let pipeline = if pipeline
+                .steps
+                .iter()
+                .all(|id| plan.steps[*id].kind.is_transform())
+            {
+                debug_assert_eq!(
+                    first_step_inputs.len(),
+                    1,
+                    "Transforms should have a single input"
+                );
+                let first_step_input_id = first_step_inputs[0];
+
+                // If all of the steps are transforms, then we use the transform pipeline.
+                executor.add_transform_pipeline(
+                    first_step_input_id,
+                    &plan,
+                    &pipeline.steps,
+                    sink,
+                )?
+            } else {
+                assert_eq!(
+                    pipeline.steps.len(),
+                    1,
+                    "Non-transforms should be in separate pipelines"
+                );
+
+                let step = &plan.steps[pipeline.steps[0]];
+                if let Some(pipeline) =
```

Comment:
Theoretically, we could instead add a "noop" or pass-through pipeline for the source. This would have the benefit of every pipeline in the plan having a corresponding executable pipeline. It could also make it easier to add logic that hapens within the worker threads for sources in the future (and/or have a "pull based source model", for instance)...

Reply:
I think that's a good idea. Though, in that case, would it make sense to just have a full `ReadPipeline` with minimal logic?

Reply:
I think if the pipeline could access the tokio runtime handle, it could actually be responsible for launching the tokio task. This would give pipelines (in general) a nice ability to actually execute async tasks, which could be beneficial.

Reply:
Yeah `ReadPipeline` would be a `no-op`, would just be there for readability if we want that.

Reply:
Filed https://github.com/kaskada-ai/kaskada/issues/814